### PR TITLE
Introducing a mechanism to inject scoped dictionary changes around pu…

### PR DESCRIPTION
**Pull request by Yoric**
…re data fragments of JavaScript.

### CODE REVIEW

File: crates/binjs_es6/src/sublanguages.rs [hunk 1]
```diff
@@ -0,0 +1,426 @@
...
+
+/// The name of the dictionary used to represent pure data subtrees (aka "JSON").
+pub const DICTIONARY_NAME_PURE_DATA: &'static str = "pojo";
+
+/// Properties observed when examining a subtree.
+#[derive(Clone, Copy, Debug)]
+enum Properties {
+    /// The subtree contains only pure data nodes.
+    PureDataSoFar {
+        /// The number of nodes in the subtree.
```
Comment (arai-a) [inline crates/binjs_es6/src/sublanguages.rs]: can you add comment what this is used for?

File: crates/binjs_es6/src/sublanguages.rs [hunk 2]
```diff
@@ -0,0 +1,426 @@
...
+/// `complex`, so we do not attempt to rewrite `pure_1` to inject a scoped
+/// dictionary change around this child. On the other hand, as we walk `pure_2`
+/// after `complex`, we already have the information, so we may decide to inject
+/// a scoped dictionary change around `pure_2`.
+///
+/// Future versions may fix this limitation.
+pub struct InjectVisitor {
+    /// If a subtree `T` is pure data and has a size >= `pure_data_threshold`, measured
+    /// in number of nodes, and if its parent is not pure data, we should rewrite `T`
+    /// to wrap it ina scoped dictionary change to DICTIONARY_NAME_PURE_DATA.
```
Comment (arai-a) [inline crates/binjs_es6/src/sublanguages.rs]: "in a"

File: crates/binjs_es6/src/sublanguages.rs [hunk 3]
```diff
@@ -0,0 +1,426 @@
...
+    }
+
+    /// The properties of the current node.
+    fn properties(&self) -> Properties {
+        self.stack.borrow().last().unwrap().clone()
+    }
+
+    /// The properties of the parent node, or `None` if there is no parent.
+    fn parent_properties(&self) -> Option<Properties> {
+        let borrow = self.stack.borrow();
```
Comment (arai-a) [inline crates/binjs_es6/src/sublanguages.rs]: can this be renamed to something more descriptive?
IMO, `borrow` doesn't explain any about what it is.
can it be `stack` or something?

File: crates/binjs_es6/src/sublanguages.rs [hunk 4]
```diff
@@ -0,0 +1,426 @@
...
+        // No need to visit the children.
+        Ok(VisitMe::DoneHere)
+    }
+
+    fn enter_spread_element_or_expression<'a>(
+        &mut self,
+        path: &WalkPath,
+        _node: &mut ViewMutSpreadElementOrExpression<'a>,
+    ) -> EnterResult {
+        // Pure data so far.
```
Comment (arai-a) [inline crates/binjs_es6/src/sublanguages.rs]: is SpreadElement pure data ?

File: crates/binjs_es6/src/sublanguages.rs [hunk 5]
```diff
@@ -0,0 +1,426 @@
...
+        return Ok(VisitMe::HoldThis(PropertiesGuard::pure_data(self, 1, path)));
+    }
+
+    fn enter_object_expression(
+        &mut self,
+        path: &WalkPath,
+        _node: &mut ObjectExpression,
+    ) -> EnterResult {
+        // Pure data so far.
+        return Ok(VisitMe::HoldThis(PropertiesGuard::pure_data(self, 1, path)));
```
Comment (arai-a) [inline crates/binjs_es6/src/sublanguages.rs]: shouldn't we handle each `ObjectProperty` items ?
are ShorthandProperty and ShorthandProperty pure data?

File: crates/binjs_es6/src/sublanguages.rs (hunk 4 above)
Comment (Yoric) [reply crates/binjs_es6/src/sublanguages.rs]: No, it isn't.

File: crates/binjs_es6/src/sublanguages.rs (hunk 5 above)
Comment (Yoric) [reply crates/binjs_es6/src/sublanguages.rs]: Good catch for `ObjectProperty`.

**Pull request merged**